consensus/polybft/extra.go

```diff
@@ -142,14 +142,81 @@ func (i *Extra) UnmarshalRLPWith(v *fastrlp.Value) error {
 	return nil
 }
 
-// ValidateBasic contains extra data basic set of validations
-func (i *Extra) ValidateBasic(parentExtra *Extra) error {
+// ValidateFinalizedHeader contains extra data validations for finalized headers
+func (i *Extra) ValidateFinalizedHeader(header *types.Header, parent *types.Header, parents []*types.Header,
+	chainID uint64, consensusBackend polybftBackend, logger hclog.Logger) error {
+	// validate committed signatures
+	blockNumber := header.Number
+	if i.Committed == nil {
+		return fmt.Errorf("failed to verify signatures for block %d, because signatures are not present", blockNumber)
+	}
+
+	if i.Checkpoint == nil {
+		return fmt.Errorf("failed to verify signatures for block %d, because checkpoint data are not present", blockNumber)
+	}
+
+	checkpointHash, err := i.Checkpoint.Hash(chainID, header.Number, header.Hash)
+	if err != nil {
+		return fmt.Errorf("failed to calculate proposal hash: %w", err)
+	}
+
+	validators, err := consensusBackend.GetValidators(blockNumber-1, parents)
+	if err != nil {
+		return fmt.Errorf("failed to validate header for block %d. could not retrieve block validators:%w", blockNumber, err)
+	}
+
+	if err := i.Committed.Verify(validators, checkpointHash, logger); err != nil {
+		return fmt.Errorf("failed to verify signatures for block %d (proposal hash %s): %w",
+			blockNumber, checkpointHash, err)
+	}
+
+	parentExtra, err := GetIbftExtra(parent.ExtraData)
+	if err != nil {
+		return fmt.Errorf("failed to verify signatures for block %d: %w", blockNumber, err)
+	}
+
+	// validate parent signatures
+	if err := i.ValidateParentSignatures(blockNumber, consensusBackend, parents,
+		parent, parentExtra, chainID, logger); err != nil {
+		return err
+	}
+
 	return i.Checkpoint.ValidateBasic(parentExtra.Checkpoint)
 }
 
-// Validate contains extra data validation logic
-func (i *Extra) Validate(parentExtra *Extra, currentValidators AccountSet, nextValidators AccountSet) error {
-	return i.Checkpoint.Validate(parentExtra.Checkpoint, currentValidators, nextValidators)
+// ValidateParentSignatures validates signatures for parent block
+func (i *Extra) ValidateParentSignatures(blockNumber uint64, consensusBackend polybftBackend, parents []*types.Header,
+	parent *types.Header, parentExtra *Extra, chainID uint64, logger hclog.Logger) error {
+	// skip block 1 because genesis does not have committed signatures
+	if blockNumber <= 1 {
```

Comment:
Does 1 genesis have signatures?

Reply:
No, the genesis block contains no signatures. Since we are validating finalized/parent signatures (namely validating signatures for block N-1 when validating block N), we skip genesis and block#1 validation. We are going to validate signatures for block#1 when validating block#2.